File: src/main/java/com/suhailkandanur/eopi/util/BinarySearchTree.java
package com.suhailkandanur.eopi.util;

/**
 * Created by suhail on 2017-01-01.
 */
public interface BinarySearchTree {
    static TreeNode constructBST(int[] array) {
        return constructBST(array, false);
    }

    static TreeNode constructBST(int[] array, boolean withParentPtr) {
        TreeNode root = null;
        for (int i = 0; array != null && i < array.length; i++) {
            int element = array[i];
            TreeNode newNode = new TreeNode(element);
            if(root == null) {
                root = newNode;
                continue;
            }
            TreeNode ptr = root;
            while(ptr != null) {
                int data = ptr.getData();
                if(element <= data) {
                    if(ptr.getLeft() == null) {
                        ptr.setLeft(newNode);
                        if(withParentPtr) {
                            newNode.setParent(ptr);
                        }
                        break;
                    } else {
                        ptr = ptr.getLeft();
                    }
                } else {
                    if(ptr.getRight() == null) {
                        ptr.setRight(newNode);
                        if(withParentPtr) {
                            newNode.setParent(ptr);
                        }
                        break;
                    } else {
                        ptr = ptr.getRight();
                    }
                }
            }
        }
        return root;
    }

    static TreeNode constructBSTwithParentPtr(int[] array) {
        return constructBST(array, true);
    }

    static void traverseInOrderBST(TreeNode root) {
        traverseInOrderBSTHelper(root);
        System.out.println();
    }

    static void traverseInOrderBSTHelper(TreeNode root) {
        if(root == null) {
            return;
        }
        traverseInOrderBSTHelper(root.getLeft());
        System.out.print(root.getData()+ ", ");
        traverseInOrderBSTHelper(root.getRight());
    }

    static  TreeNode findElement(TreeNode root, int element) {
        TreeNode ptr = root;
        while(ptr != null) {
            int data = ptr.getData();
            if(data == element) {
                break;
            } else if(element < data) {
                ptr = ptr.getLeft();
            } else {
                ptr = ptr.getRight();
            }
        }
        return ptr;
    }
}

File: python/BinarySearchTree/binarysearchtree.py
from treenode import TreeNode


class BinarySearchTree:
    def __init__(self):
        self.init = 1
        self.root = None

    def add(self, element):
        self.root = self.addHelper(self.root, element)
        return self.root

    def addHelper(self, root, element):
        if root is None:
            node = TreeNode(element)
            return node

        if root.getData() >= element:
            root.setLeft(self.addHelper(root.getLeft(), element))
        else:
            root.setRight(self.addHelper(root.getRight(), element))

        return root

    def find(self, element):
        return self.findHelper(self.root, element)

    def findHelper(self, root, element):
        if root is None:
            return None

        if root.getData() == element:
            return root
        elif root.getData() > element:
            return self.findHelper(root.getLeft(), element)
        else:
            return self.findHelper(root.getRight(), element)

    def keyLargerThanK(self, k):
        return self.keyLargerThanKHelper(self.root, k)

    def keyLargerThanKHelper(self, root, k):
        if root is None:
            return None

        if root.getData() <= k:
            node = self.keyLargerThanKHelper(root.getRight(), k)
            if node is None or node.getData() == k:
                return None
            return node
        else:
            node = self.keyLargerThanKHelper(root.getLeft(), k)
            if node is None:
                return root
            return node
